analyzer/src/main/kotlin/PackageCurationProvider.kt

```diff
@@ -64,10 +65,10 @@ fun interface PackageCurationProvider {
     }
 
     /**
-     * Return all available [PackageCuration]s for the provided [pkgIds], associated by the package's [Identifier]. Each
-     * list of curations must be non-empty; if no curation is available for a package, the returned map must not contain
-     * a key for that package's identifier at all.
+     * Return all available [PackageCuration]s for the provided [packages], associated by the package's [Identifier].
+     * Each list of curations must be non-empty; if no curation is available for a package, the returned map must not
+     * contain a key for that package's identifier at all.
      */
     // TODO: Maybe make this a suspend function, then all implementing classes could deal with coroutines more easily.
-    fun getCurationsFor(pkgIds: Collection<Identifier>): Map<Identifier, List<PackageCuration>>
+    fun getCurationsFor(packages: Collection<Package>): Map<Identifier, List<PackageCuration>>
```

Comment:
I would prefer an API that is more tailored for the use case. Packages contain a lot more data than required to lookup curations which leads to ugly constructs when using the API like `Package.EMPTY.copy(id = id)` in the tests. If the `Identifier` is not sufficient a dedicated request object would be better, for example if the `id` and `binaryArtifact` were required:

```kotlin
data class PackageCurationQuery(
  id: Identifier,
  binaryArtifact: RemoteArtifact = RemoteArtifact.EMPTY
)
```

Reply:
At first, I also didn't like so much to pass in "heavyweight" `Package` instances. But actually, most "real" code (not counting tests) becomes simpler this way, as existing collections of packages do not need to be mapped to ids, avoiding the creation of temporary collections, or of temporary helper classes like your proposed `PackageCurationQuery`.

Reply:
This could be solved by having an additional `getCurationsFor(packages: Collection<Package>)` function in `PackageCurationProvider` that does the mapping for packages and calls the abstract `getCurationsFor` function while still ensuring that implementations of `PackageCurationProvider` only get the relevant data.

Reply:
For providers, like ClearlyDefined, that actually require the full package, how would you then prevent callers from calling `fun getCurationsFor(pkgIds: Collection<Identifier>)` instead of `fun getCurationsFor(packages: Collection<Package>)`?

Reply:
What could work is to turn `PackageCurationProvider` into an `abstract class` with:

```
    fun getCurationsFor(packages: Collection<Package>): Map<Identifier, List<PackageCuration>> =
        getCurationsFor(packages.map { it.id })

    protected fun getCurationsFor(ids: Collection<Identifier>): Map<Identifier, List<PackageCuration>> =
        throw NotImplementedError()

```

Reply:
I didn't mean to keep the function that takes only ids, but that we should carefully extend the API to not include values that are not required. What I want to avoid is a situation like the one fixed [here](https://github.com/oss-review-toolkit/ort/pull/6366/commits/b1cec227e1264ee8abbe0f413ac9469f93c65c09) where the API required far more information than is actually required which unnecessarily complicates usage and testing.

So the proposal was to define which values are actually required to look up curation and limit the API to these, and to have a convenience function to extract that data from packages, for example:

```kotlin
data class PackageCurationQuery(
  val id: Identifier,
  val binaryArtifact: RemoteArtifact,
  val vcs: VcsInfo
)

// Curation providers only need to implement this function.
fun getCurationsFor(queries: Set<PackageCurationQuery>): Map<PackageCurationQuery, List<PackageCuration>>

fun getCurationsFor(packages: Set<Package>) = getCurationsFor(packages.map { it.toPackageCurationQuery() })

fun Package.toPackageCurationQuery() = PackageCurationQuery(id, binaryArtifact, vcsProcessed)
```

> For providers, like ClearlyDefined or also https://github.com/oss-review-toolkit/ort/issues/6389, that actually require the full package when implemented correctly,

If this is the case then the API must take full packages, but is it? Do ClearlyDefined or OSSelot require for example the `authors`, `description` or `isMetadataOnly` properties?

Reply:
> If this is the case then the API must take full packages, but is it? Do ClearlyDefined or OSSelot require for example the `authors`, `description` or `isMetadataOnly` properties?

No, admittedly they don't. To me the question simply is where to draw the line: Starting with which percentage of the original properties required should we stop introducing a dedicated `Query` data class?

With [your change](https://github.com/oss-review-toolkit/ort/pull/6366/commits/b1cec227e1264ee8abbe0f413ac9469f93c65c09) advisors now operate on packages, [scanners can operate on (a single) package(s)](https://github.com/oss-review-toolkit/ort/blob/ed1be558ad222b67aa0ffee1a00fe1f6758c0045/scanner/src/main/kotlin/ScannerWrapper.kt#L72), so also from that perspective it makes sense to me if curation providers would operate on packages, too.

Reply:
I think it would be difficult to draw a hard line but in this case it seemed to make sense to me to have a dedicated class because only a few properties would be needed in addition to the id and most package properties would not be required. I wouldn't compare the curation providers the advisor or scanner because these are top-level tools and the curation providers have a very specific use case.

What's actually your concern, is it that you don't want to introduce another class?

Reply:
> is it that you don't want to introduce another class?

Exactly. I'm reluctant to introduce such small helper data classes (similar also for return values) as more classes simply make it harder to get an overview of class hierarchies and the code base in general.

Also the benefit of having such a class is unclear to me in this case: If we anyway have `fun getCurationsFor(packages: Set<Package>)`, which callers (currently) would call `fun getCurationsFor(queries: Set<PackageCurationQuery>)`? And: Is it really that bad to give advisors access to the rest of the (currently unneeded) `Package` properties, when in exchange it avoids a mapping?

Reply:
The benefit of a dedicated class to me is a bettter specification of the package curation provider API which also leads to cleaner tests. For example, with `Package.EMPTY.copy(id = id)` it's less clear which properties are actually used by the providers. Let's talk about this in the dev meeting today.

Reply:
For reference, it was agreed in the developer meeting to keep the API, for example because future provider implementations might need access to other package properties.